proposals/0000-midstream-bindists.md

```diff
@@ -0,0 +1,199 @@
+
+* https://gitlab.haskell.org/ghc/ghcup-metadata/-/issues/2
+* https://gitlab.haskell.org/ghc/ghc/-/issues/24000
+
+The breakage was left unattended and some bindists completely vanished from gitlab CI artifacts, because of misconfiguration.
+Here's a graph of nightlies availability: https://grafana.gitlab.haskell.org/d/ab109e66-a8a1-4ae9-b976-40e2dfe281ab/availabilitie-of-ghc-nightlies-via-ghcup?orgId=2
+
+## Prior Art and Related Efforts
+
+So far, GHCup developers have tried to close the gap, doing signifcant work on tooling CIs and building bindists manually.
+
+Bindists for cabal-install are now produced by GHCup's own CI: https://github.com/haskell/ghcup-metadata/blob/develop/.github/workflows/cabal-release.yaml
+
+HLS will likely follow shortly.
+
+## Technical Content
+
+We propose here to create a joint project of "installation experience" developers to get funding and maintain
+all tooling bindists centrally and autonomically.
+
+This will allow:
+
+* greater quality assurance
+* more unofficial platform support (e.g. FreeBSD, i386, ...)
+* more bindists for uncommon linux distros
+* continuously updated and maintained bindists (including historical GHC versions)
+* actually working nightlies
+
+One major concern that was brought up about building bindists midstream was if/how the testing will be done.
+Building bindists midstream is a great opportunity to put more pressure on e.g. GHC HQ to fix issues with the test bindist:
+
+* https://gitlab.haskell.org/ghc/ghc/-/issues/22726
+* https://gitlab.haskell.org/ghc/ghc/-/issues/22723
+* https://gitlab.haskell.org/ghc/ghc/-/issues/22727
+
+The idea is that bindists should be primarily tested **on the users system**, because that is where they're going to run.
+It is great to know that e.g. the test suite passes on GHC CI, but that may have little value in different environments.
+
+Additionally, issues with tests can flow back to upstream developers and we may develop workflows and processes to streamline
+this type of feedback. Early release candidates can assist with this workflow.
+
+Another effort closely related to this project is that upstream projects should maintain
+a more end-user focussed build system for producing release binaries, making it easier
+for both distributors and end-user to build such release binaries correctly themselves.
+Related issues:
+
+* https://github.com/haskell/cabal/issues/9461
+* https://github.com/haskell/haskell-language-server/issues/3878
+
+GHCup will also require to implement revisions to make updated bindists feasible: https://github.com/haskell/ghcup-hs/issues/361
+
+## Timeline
+
+* 3 months: proof of concept of a central GitHub CI building the entire toolchain
+* 6 months: support for uncommon platforms, bindist testing workflow, automated integration with GHCup, etc.
+* 9 months: nightlies working for GHC and cabal
+
+## Budget
+
+This might require a full time developer for at least half a year,
+as well as help from volunteers.
```

Comment:
What would be the costs once the initial setup is working?

Reply:
It is continuous effort, so ultimately yes.

Reply:
Good, then this should be in the budget section.